### PR TITLE
3.0 major update

### PR DESCRIPTION
**Pull request by appleneko2001**
> This is a PR draft, it will be able to merge after all related commits are applied.

# 3.0 Feature update PR

## What this PR used for?
This PR will making biggest changes on dev-branch to make library better looks, performances, fixes, improve code formatting or etc. The `Material.Avalonia` library 3.0 stable release will be almost done after apply this PR.

UPDATE: This PR allows you to use Avalonia 11.0, it closes issue such as #180 

## Which changes I made in this PR?
- I redesigned most controls to minimalize templates variant and stylize those templates instead and improve on performances. For example, minimalize templates of TextBox as one template will take less lines on source code, prevent performance hits while changing templates frequently, which making library more better and more sufficient on performances. 
- I corrected queues (resort) of style include list to solve problem like "applied styles, but not as expected". For example, we want to apply default TextBlock styles and before it had MenuItem styles which include styling inner texts of each menu items. This problem will make text of menu item styled as TextBlock style want to and it looks will be weird.
- I corrected Material.Dialog mechanism of works to make it create an control and it will be styled on everywhere. It means you can take out the main control of created dialog by `DialogHelper` and use it somewhere like `DialogHost` library without any critical problem.
- ~~I corrected classes definitions on almost all controls to make them follows most conventions of CSS naming rules. That means you can forgot about upper-case and use lower-case classes naming. Well that doesnt mean we'll remove old classes definitions, they will be kept until next major update or incompatibility on fixes or smth like that. Please get ready to use newer classes naming before they stops working.~~ **(Those changes will be applied to compatibility classes set because I'm gonna use newest way to apply theme to controls, it is included in this PR. The classes set can be used via StyleInclude)**
- I removed features table in Material.Demo, because I didnt update it after changes and it appears might not such important. I know this changes is not cool for some ppl, but I think there should be better way than this thing.
- I corrected something related with Material.Colors, which they used for pick most contrast color for texts and icons. It should fixes problem like "text or icons in buttons is hard to see", "ugly coloring of their inner text or icons of buttons" or anything related with coloring issues.
- I used new technique in our library named "Static naming collections". It used for prevent situations like "multiple strings but same thing". That means I trying to fix memory leaks on string instances by making templates of all controls to use one string instance on name property. Plus, its will be easier to reference names on their controls in IDE because they all are static property in one static class.
- I moved all controls C# code to `Controls` folder to make source code folder more cleaner and easy to find their source code. 
- they are too much I cant remember them all sorry((

### It fixes or solve issues like:
- Fix: not all properties is usable with our library.
- Fix: uncomfortable classes naming
- Fix: styles applied not as expected or appears other glitches.
- Fix: Material.Dialog library, correct how it should be work to solve issue like #164 .
- Fix: Filled TextBox rendering glitch
- Fix: Outline TextBox clipping abnormally problem
- Fix: `ToggleSwitch` knobs is undraggable.
- Fix: bad color contrast in inner text / icon of buttons.
- Minimalize memory leaks on their templates
- Fix: IME not working in Material.Demo
> Tips: To fix this issue in your avaloniaUI applications is not hard. Use STA (single threaded apartment) on Main entry of your applications, and make sure your property "InputMethod.IsInputMethodEnabled" on each text boxes are enabled, or you can refer how we did it.
- Fix: not all characters of text block or text boxes is rendered.
- Fix: comboboxes is unstyled.
- Fix: circular progress indicator had bad performance.
- Fix: TreeView is bad styled.

### It improve looks of next controls:
- Button (included FAB)
- ToggleSwitch
- TextBox
- TextBlock
- Slider
- CheckBox
- RadioButton
- ComboBox
- ColorZone
- Separator
- TreeView
- TimePicker, DatePicker
- ScrollViewer

### **UPDATE: Add new features:**
+ **Compatibility classes set (you may include it, if you want apply theme by class)**

### and It might breaks next things:
- old classes naming definition (definitely, if you dont use classes set.)
- some assist class will be deprecated and removed, or removes partial things of them
- use Material.Styles.Controls namespace instead of Material.Styles if you use our controls in your applications, in code-behind.
- customized styling to replace partial things from our libs in your applications.

## TL;DR:
Fix almost everything (including styles, controls) not works if possible, add some new features, use better classes naming, remove some bad APIs and optimize styles. The Dialog library will be fixed, too. Meanwhile the old API or some features could be broken or unstable.

### CODE REVIEW

**Comment by SKProCH** [comment]
Good job! I'll see the changes up close in the near future

File: Material.Styles/Themes/MaterialThemeBase.cs [hunk 1]
```diff
@@ -39,34 +39,71 @@ public MaterialThemeBase(IServiceProvider serviceProvider)
 
         private IResourceDictionary LoadedResourceDictionary => (Loaded as Avalonia.Styling.Styles)!.Resources;
 
-        public static readonly DirectProperty<MaterialThemeBase, ITheme> CurrentThemeProperty =
-            AvaloniaProperty.RegisterDirect<MaterialThemeBase, ITheme>(
+        public static readonly DirectProperty<MaterialThemeBase, ITheme?> CurrentThemeProperty =
```
Comment (SKProCH) [inline Material.Styles/Themes/MaterialThemeBase.cs:42]: It is assumed that CurrentTheme always has a value

File: Material.Styles/Themes/MaterialThemeBase.cs [hunk 2]
```diff
@@ -39,34 +39,71 @@ public MaterialThemeBase(IServiceProvider serviceProvider)
...
+                
+                    _currentTheme = newTheme;
+                    RaisePropertyChanged(CurrentThemeProperty, oldTheme, newTheme);
+                    StartUpdatingTheme(oldTheme, newTheme);
+                }
+                else
+                {
+                    _currentTheme = null;
+                    RaisePropertyChanged(CurrentThemeProperty, oldTheme, null);
+                    StartUpdatingTheme(oldTheme, null);
```
Comment (SKProCH) [inline Material.Styles/Themes/MaterialThemeBase.cs:85]: We can't assign `null` to `CurrentTheme`, cuz it cause `NullReferenceException` here. We don't have to support null here. Or, at the very least, I can't find a use for it.

File: Material.Styles/Themes/MaterialThemeBase.cs [hunk 3]
```diff
@@ -39,34 +39,71 @@ public MaterialThemeBase(IServiceProvider serviceProvider)
...
             }
         }
 
-        public IObservable<ITheme> CurrentThemeChanged => this.GetObservable(CurrentThemeProperty);
+        public IObservable<ITheme?> CurrentThemeChanged => this.GetObservable(CurrentThemeProperty);
+        
+        /// <summary>
+        /// This event is raised when all brushes is changed.
+        /// </summary>
+        public event EventHandler? ThemeChanged;
```
Comment (SKProCH) [inline Material.Styles/Themes/MaterialThemeBase.cs:95]: Hmmm, this event is interesting idea

File: Material.Styles/Themes/MaterialThemeBase.cs [hunk 4]
```diff
@@ -130,19 +168,20 @@ public MaterialThemeBase(IServiceProvider serviceProvider)
             }
         });
 
-        private Task UpdateThemeAsync(ITheme? oldTheme, ITheme newTheme) {
-            return Task.WhenAll(
+        private Task UpdateThemeAsync(ITheme? oldTheme, ITheme newTheme)
```
Comment (SKProCH) [inline Material.Styles/Themes/MaterialThemeBase.cs:171]: I've updated theme applying in dev branch

**Review by SKProCH — COMMENTED** [review]
I think changing namespaces for controls is a good solution from an architectural point of view, but it's also a terrible solution from a backward compatibility point of view. Users who want to update the version in their projects will have to change the declaration of each of our controls in their code. Considering that VS has no hints for this it would be terrible.

File: Material.Demo/Pages/SlidersDemo.axaml [hunk 5]
```diff
@@ -3,131 +3,257 @@
...
                     </showMeTheXaml:XamlDisplay>
                     <TextBlock />
                 </StackPanel>
-                <StackPanel Classes="Content" Orientation="Horizontal">
-                    <showMeTheXaml:XamlDisplay UniqueId="BubbleSlider3">
+                <StackPanel Classes="Content" Orientation="Horizontal" >
+                    <showMeTheXaml:XamlDisplay UniqueId="BubbleSlider3" >
                         <Slider IsSnapToTickEnabled="True"
-                                Classes="Modern Discrete Accent Ticks"
+                                Classes=" Discrete Accent Ticks"
```
[the remark this answers is not in the document]
Comment (appleneko2001) [reply Material.Demo/Pages/SlidersDemo.axaml:128]: you can give some idea for new classes naming
but if you asking why I dont change their classes in demo, because I want do it as the last commit of this PR
plus I want keep most old classes naming until next major update to give others prepare to update.

File: Material.Dialog/Views/TextFieldDialog.axaml.cs [hunk 6]
```diff
@@ -3,6 +3,7 @@
 using Material.Dialog.Interfaces;
 using Material.Dialog.ViewModels;
 using System;
+using Avalonia;
```
[the remark this answers is not in the document]
Comment (appleneko2001) [reply Material.Dialog/Views/TextFieldDialog.axaml.cs:6]: wait let me do code cleaning

File: Material.Styles/ColorZone.xaml [hunk 7]
```diff
@@ -2,131 +2,47 @@
...
-        <Setter Property="Foreground" Value="{DynamicResource MaterialDesignDarkForeground}" />
-    </Style>
-    
-    <Style Selector="cc|ColorZone[Mode=Dark] Button,
-                     cc|ColorZone[Mode=Dark] ToggleButton">
-        <Setter Property="Foreground" Value="{DynamicResource MaterialDesignDarkForeground}" />
-        <Setter Property="assists:ToggleButtonAssist.UncheckedForeground" Value="{DynamicResource MaterialDesignDarkForeground}"/>
-    </Style>
-
+<!-- Compatibility issue: require preloaded MaterialTheme
```
[the remark this answers is not in the document]
Comment (appleneko2001) [reply Material.Styles/ColorZone.xaml:37]: I was trying some new technique to provide background and foreground brushes, because the style way is not works sometimes. I could rollback those changes if you dont like the newer ColorZone

File: Material.Styles/Material.Styles.csproj [hunk 8]
```diff
@@ -59,4 +59,8 @@
       <PackageReference Include="DialogHost.Avalonia" Version="0.4.0" />
     </ItemGroup>
 
+    <ItemGroup>
```
[the remark this answers is not in the document]
Comment (appleneko2001) [reply Material.Styles/Material.Styles.csproj:62]: yes, there was a source code named MaterialCalendar.cs, which I want to correct default Calendar control behaviour but it is empty file. There was a problem with Calendar control, it doesnt support set special classes for multiselected items.

File: Material.Styles/Themes/MaterialThemeBase.cs (hunk 4 above)
Comment (appleneko2001) [reply Material.Styles/Themes/MaterialThemeBase.cs:171]: wait let me do some changes for this code

File: Material.Styles/Themes/MaterialThemeBase.cs (hunk 1 above)
Comment (appleneko2001) [reply Material.Styles/Themes/MaterialThemeBase.cs:42]: I pushed some commits, including remove nullable annotations

File: Material.Styles/ButtonSpinner.xaml [hunk 9]
```diff
@@ -1,11 +1,14 @@
 <Styles xmlns="https://github.com/avaloniaui"
         xmlns:assists="clr-namespace:Material.Styles.Assists">
     <Style Selector="ButtonSpinner">
-        <Setter Property="BorderThickness" Value="1" /> <!-- FixBorderThickness -->
+        <Setter Property="BorderThickness" Value="0" />
         <Setter Property="HorizontalContentAlignment" Value="Stretch" />
         <Setter Property="VerticalContentAlignment" Value="Center" />
         <Setter Property="Foreground" Value="{DynamicResource PrimaryHueMidForegroundBrush}"/>
         <Setter Property="BorderBrush" Value="{DynamicResource MaterialDesignTextAreaBorder}" />
+        <Setter Property="Template">
```
[the remark this answers is not in the document]
Comment (appleneko2001) [reply Material.Styles/ButtonSpinner.xaml:9]: it was empty because I didnt redesign that control, I might fix it today

File: Material.Styles/Button.xaml [hunk 10]
```diff
@@ -1,246 +1,302 @@
...
+        <Setter Property="Padding" Value="12" />
+        <Setter Property="Height" Value="32" />
+        <Setter Property="Width" Value="32" />
+        <Setter Property="assists:ShadowAssist.ShadowDepth" Value="Depth0" />
+        <Setter Property="CornerRadius" Value="100000"/>
+    </Style>
+
+    <!-- Outlined buttons -->
+
+    <Style Selector="Button.Outline" >
```
[the remark this answers is not in the document]
Comment (appleneko2001) [reply Material.Styles/Button.xaml:272]: yep, for compatibility , but we will remove them at next major update

File: Material.Demo/Material.Demo.csproj [hunk 11]
```diff
@@ -6,10 +6,10 @@
     <Nullable>enable</Nullable>
   </PropertyGroup>
   <ItemGroup>
-    <PackageReference Include="Avalonia" Version="0.10.11" />
-    <PackageReference Include="Avalonia.Controls.DataGrid" Version="0.10.11" />
-    <PackageReference Include="Avalonia.Desktop" Version="0.10.11" />
-    <PackageReference Include="Avalonia.Diagnostics" Version="0.10.11" />
+    <PackageReference Include="Avalonia" Version="0.10.14" />
```
[the remark this answers is not in the document]
Comment (appleneko2001) [reply Material.Demo/Material.Demo.csproj:9]: ok

File: Material.Demo/Pages/SlidersDemo.axaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (SKProCH) [reply Material.Demo/Pages/SlidersDemo.axaml:128]: Actually 3.0 - is a major update already, isn't it?

File: Material.Styles/ColorZone.xaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (SKProCH) [reply Material.Styles/ColorZone.xaml:37]: I need to see more details, I'll look at it later.

**Comment by appleneko2001** [comment]
> @appleneko2001 but actually it works for demo, isn't it?

my ide always show it as error, sometimes it blocks me to compile the demo

File: Material.Styles/ColorZone.xaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (appleneko2001) [reply Material.Styles/ColorZone.xaml:37]: please let me know this techniqhe is helpful or not thx

File: Material.Avalonia/Material.Avalonia.csproj [hunk 12]
```diff
@@ -41,7 +41,7 @@
     </ItemGroup>
 
     <ItemGroup>
-      <PackageReference Include="Avalonia" Version="0.10.15" />
+      <PackageReference Include="Avalonia" Version="0.10.16" />
```
Comment (SKProCH) [inline Material.Avalonia/Material.Avalonia.csproj:44]: I think Material.Avalonia should depend on the earliest version of Avalonia possible, so as not to force users to update the Avalonia version in their projects. 
Conversely, Material.Demo should be based on the latest version of Avalonia, to keep track of changes to Avalonia and possibly find bugs that will show up with an Avalonia update.

**Comment by JanTamis** [comment]
When can we expect this release?

**Comment by SKProCH** [comment]
Hello, @JanTamis 

Currently we have [3.0.0-avalonia10-preview2](https://www.nuget.org/packages/Material.Avalonia/3.0.0-avalonia10-preview2) for Avalonia 10 based on  383e88e. And soon i release preview packages for Avalonia 11 with this changes. 

As for this PR, its better ask @appleneko2001, because unfortunately i don't have time to fully support this project now. 
PR warmly welcomed.

**Comment by JanTamis** [comment]
@appleneko2001 Hi how can i help you with the 3.0 update?

File: Material.Styles/Resources/Themes/Slider.axaml
Comment (appleneko2001) [inline Material.Styles/Resources/Themes/Slider.axaml:1]: Seems like in newer AvaloniaUI deprecated old way to direct access to UI component position data "TransformedBounds", probably need something else to access

**Comment by Janek91** [comment]
Can I help somehow with this?

**Comment by SKProCH** [comment]
@Janek91, you can take one of "up for grabs" issues, actually: https://github.com/AvaloniaCommunity/Material.Avalonia/issues?q=is%3Aopen+is%3Aissue+label%3A%22up+for+grabs%22

**Comment by SKProCH** [comment]
Actually, this is branch is only branch with main package version (and Avalonia 11 target) developed currently. So, since there is most likely no releases for other branches (v2.* and v3.0-avalonia10), i want to merge it to `dev` (and merge `dev` into `master` in the future)
